File: src/pgo/model/tla/TLAVariableDeclaration.java
package pgo.model.tla;

import java.util.List;
import java.util.stream.Collectors;

import pgo.util.SourceLocation;

public class TLAVariableDeclaration extends TLAUnit {
	
	List<TLAIdentifier> variables;

	public TLAVariableDeclaration(SourceLocation location, List<TLAIdentifier> variables) {
		super(location);
		this.variables = variables;
	}
	
	@Override
	public TLAVariableDeclaration copy() {
		return new TLAVariableDeclaration(getLocation(), variables.stream().map(TLAIdentifier::copy).collect(Collectors.toList()));
	}

	public List<TLAIdentifier> getVariables() {
		return variables;
	}

	@Override
	public <T, E extends Throwable> T accept(TLAUnitVisitor<T, E> v) throws E {
		return v.visit(this);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((variables == null) ? 0 : variables.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TLAVariableDeclaration other = (TLAVariableDeclaration) obj;
		if (variables == null) {
			return other.variables == null;
		} else return variables.equals(other.variables);
	}

}

File: src/pgo/model/tla/TLAVariableDeclaration.scala
package pgo.model.tla

import scala.jdk.CollectionConverters._
import pgo.util.SourceLocation

final class TLAVariableDeclaration(loc: SourceLocation, val variables: List[TLAIdentifier]) extends TLAUnit(loc) with TLADefinitionComposite {
  def this(loc: SourceLocation, variables: java.util.List[TLAIdentifier]) =
    this(loc, variables.asScala.toList)

  def getVariables: java.util.List[TLAIdentifier] = variables.asJava

  override val definitions: List[TLADefinition] = List(this)

  override def copy: TLAVariableDeclaration = new TLAVariableDeclaration(loc, variables.map(_.copy()))

  override val members: List[TLADefinition] = variables

  override def accept[T, E <: Throwable](v: TLAUnitVisitor[T, E]): T = v.visit(this)

  override def equals(other: Any): Boolean = other match {
    case that: TLAVariableDeclaration =>
      variables == that.variables
    case _ => false
  }

  override def hashCode(): Int = variables.hashCode

  override def toString: String = s"VARIABLES ${variables.mkString(",")}"
}
